diff --git a/adb/daemon/main.cpp b/adb/daemon/main.cpp
index 4721e2f..266200c 100644
--- a/adb/daemon/main.cpp
+++ b/adb/daemon/main.cpp
@@ -143,11 +143,13 @@ static void drop_privileges(int server_port) {
         // minijail_enter() will abort if any priv-dropping step fails.
         minijail_enter(jail.get());

+       /*
         if (root_seclabel != nullptr) {
             if (selinux_android_setcon(root_seclabel) < 0) {
                 LOG(FATAL) << "Could not set SELinux context";
             }
-        }
+        }*/
+
         std::string error;
         std::string local_name =
             android::base::StringPrintf("tcp:%d", server_port);
diff --git a/adb/sockets.cpp b/adb/sockets.cpp
index 63b7df6..334cb57 100644
--- a/adb/sockets.cpp
+++ b/adb/sockets.cpp
@@ -387,6 +387,7 @@ asocket* create_local_socket(int fd) {
     if (s == NULL) {
         fatal("cannot allocate socket");
     }
+    disable_tcp_nagle(fd);
     s->fd = fd;
     s->enqueue = local_socket_enqueue;
     s->ready = local_socket_ready;
diff --git a/adb/transport_local.cpp b/adb/transport_local.cpp
index 5a0089f..99769e5 100644
--- a/adb/transport_local.cpp
+++ b/adb/transport_local.cpp
@@ -173,7 +173,8 @@ static void server_socket_thread(void* arg) {
             D("server: new connection on fd %d", fd);
             close_on_exec(fd);
             disable_tcp_nagle(fd);
-            register_socket_transport(fd, "host", port, 1);
+            //register_socket_transport(fd, "host", port, 1);
+           register_socket_transport(fd, NULL, port, 1);
         }
     }
     D("transport: server_socket_thread() exiting");
@@ -239,6 +240,7 @@ static void qemu_socket_thread(void* arg) {
     adb_thread_setname("qemu socket");
     D("transport: qemu_socket_thread() starting");

+    adb_thread_create(server_socket_thread, arg);
     /* adb QEMUD service connection request. */
     snprintf(con_name, sizeof(con_name), "qemud:adb:%d", port);

@@ -248,7 +250,7 @@ static void qemu_socket_thread(void* arg) {
         /* This could be an older version of the emulator, that doesn't
          * implement adb QEMUD service. Fall back to the old TCP way. */
         D("adb service is not available. Falling back to TCP socket.");
-        adb_thread_create(server_socket_thread, arg);
+        //adb_thread_create(server_socket_thread, arg);
         return;
     }

