diff --git a/rootdir/init.rc b/rootdir/init.rc
index 415d4adb6..dd7312ec7 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -298,7 +298,7 @@ on late-init
 on post-fs
     start logd
     # once everything is setup, no need to modify /
-    mount rootfs rootfs / ro remount
+    mount rootfs rootfs / rw remount
     # Mount shared so changes propagate into child namespaces
     mount rootfs rootfs / shared rec
     # Mount default storage into root namespace
