diff -uNr a/vendor/anbox/external/cpu_features/src/hwcaps.c b/vendor/anbox/external/cpu_features/src/hwcaps.c
--- a/vendor/anbox/external/cpu_features/src/hwcaps.c	2020-04-12 22:38:24.159740150 +0800
+++ b/vendor/anbox/external/cpu_features/src/hwcaps.c	2020-04-12 22:39:22.499740177 +0800
@@ -91,7 +91,8 @@
     D("Could not dlopen() C library: %s\n", dlerror());
     return 0;
   }
-  func = (getauxval_func_t*)dlsym(libc_handle, "getauxval");
+  //func = (getauxval_func_t*)dlsym(libc_handle, "getauxval");
+  *(void**)(&func) = dlsym(libc_handle, "getauxval");
   if (!func) {
     D("Could not find getauxval() in C library\n");
   } else {
diff -uNr a/vendor/anbox/products/anbox.mk b/vendor/anbox/products/anbox.mk
--- a/vendor/anbox/products/anbox.mk	2020-04-12 22:38:24.159740150 +0800
+++ b/vendor/anbox/products/anbox.mk	2020-04-12 22:39:22.499740177 +0800
@@ -54,7 +54,9 @@
 PRODUCT_CHARACTERISTICS := emulator
 
 # Include drawables for all densities
-PRODUCT_AAPT_CONFIG := normal
+PRODUCT_AAPT_CONFIG := normal xhdpi
+PRODUCT_PROPERTY_OVERRIDES += \
+    ro.sf.lcd_density=320 \
 
 PRODUCT_COPY_FILES += \
 	vendor/anbox/scripts/anbox-init.sh:root/anbox-init.sh \
@@ -94,7 +96,7 @@
 
 $(call inherit-product, $(SRC_TARGET_DIR)/product/aosp_base.mk)
 # Extend heap size we use for dalvik/art runtime
-$(call inherit-product, frameworks/native/build/tablet-10in-xhdpi-2048-dalvik-heap.mk)
+$(call inherit-product, frameworks/native/build/phone-xhdpi-2048-dalvik-heap.mk)
 
 PRODUCT_COPY_FILES += \
 	vendor/anbox/products/anbox.xml:system/etc/permissions/anbox.xml
diff -uNr a/vendor/anbox/products/arm64/BoardConfig.mk b/vendor/anbox/products/arm64/BoardConfig.mk
--- a/vendor/anbox/products/arm64/BoardConfig.mk	2020-04-12 22:38:24.159740150 +0800
+++ b/vendor/anbox/products/arm64/BoardConfig.mk	2020-04-12 22:39:22.499740177 +0800
@@ -12,11 +12,11 @@
 TARGET_CPU_VARIANT := generic
 TARGET_CPU_ABI := arm64-v8a
 
-# TARGET_2ND_ARCH := arm
-# TARGET_2ND_ARCH_VARIANT := armv7-a-neon
-# TARGET_2ND_CPU_VARIANT := cortex-a15
-# TARGET_2ND_CPU_ABI := armeabi-v7a
-# TARGET_2ND_CPU_ABI2 := armeabi
+TARGET_2ND_ARCH := arm
+TARGET_2ND_ARCH_VARIANT := armv7-a-neon
+TARGET_2ND_CPU_VARIANT := cortex-a15
+TARGET_2ND_CPU_ABI := armeabi-v7a
+TARGET_2ND_CPU_ABI2 := armeabi
 
 AUDIOSERVER_MULTILIB := 64
 
diff -uNr a/vendor/anbox/products/arm64/device.mk b/vendor/anbox/products/arm64/device.mk
--- a/vendor/anbox/products/arm64/device.mk	2020-04-12 22:38:24.159740150 +0800
+++ b/vendor/anbox/products/arm64/device.mk	2020-04-12 22:39:22.499740177 +0800
@@ -15,11 +15,11 @@
 #
 
 # Copy the 64-bit primary, 32-bit secondary zygote startup script
-PRODUCT_COPY_FILES += system/core/rootdir/init.zygote64.rc:root/init.zygote64.rc
+PRODUCT_COPY_FILES += system/core/rootdir/init.zygote64_32.rc:root/init.zygote64_32.rc
 
 # Set the zygote property to select the 64-bit primary, 32-bit secondary script
 # This line must be parsed before the one in core_minimal.mk
-PRODUCT_DEFAULT_PROPERTY_OVERRIDES += ro.zygote=zygote64
+PRODUCT_DEFAULT_PROPERTY_OVERRIDES += ro.zygote=zygote64_32
 
-TARGET_SUPPORTS_32_BIT_APPS := false
+TARGET_SUPPORTS_32_BIT_APPS := true
 TARGET_SUPPORTS_64_BIT_APPS := true
diff -uNr a/vendor/anbox/scripts/anbox-init.sh b/vendor/anbox/scripts/anbox-init.sh
--- a/vendor/anbox/scripts/anbox-init.sh	2020-04-12 22:34:14.371740035 +0800
+++ b/vendor/anbox/scripts/anbox-init.sh	2020-04-12 22:40:00.043740194 +0800
@@ -25,6 +25,7 @@
 		chown system:system /dev/$f
 		chmod 0666 /dev/$f
 	done
+	cp -rf /system/etc/* /etc/
 
 	if [ -e "/dev/tun" ] ; then
 		chown system:vpn /dev/tun
diff -uNr a/vendor/anbox/scripts/create-package.sh b/vendor/anbox/scripts/create-package.sh
--- a/vendor/anbox/scripts/create-package.sh	2020-04-12 22:38:24.159740150 +0800
+++ b/vendor/anbox/scripts/create-package.sh	2020-04-12 22:39:22.499740177 +0800
@@ -23,12 +23,19 @@
 sudo cp -ar $workdir/system/* $rootfs/system
 sudo umount $workdir/system
 
-gcc -o $workdir/uidmapshift external/nsexec/uidmapshift.c
-sudo $workdir/uidmapshift -b $rootfs 0 100000 65536
 
 # FIXME
 sudo chmod +x $rootfs/anbox-init.sh
 
+if [ -e android.img ]; then
+    DATE=`date +%F_%R`
+    SAVETO=android-old-$DATE.img
+    echo "#########################################################"
+    echo "# WARNING: Old android.img still exists.                 "
+    echo "#          Moving it to $SAVETO.                         "
+    echo "#########################################################"
+    mv android.img $SAVETO
+fi
 sudo mksquashfs $rootfs android.img -comp xz -no-xattrs
 sudo chown $USER:$USER android.img
 
