diff --git a/vendor/anbox/android/appmgr/src/org/anbox/appmgr/LauncherActivity.java b/vendor/anbox/android/appmgr/src/org/anbox/appmgr/LauncherActivity.java
index d89b03e..818f621 100644
--- a/vendor/anbox/android/appmgr/src/org/anbox/appmgr/LauncherActivity.java
+++ b/vendor/anbox/android/appmgr/src/org/anbox/appmgr/LauncherActivity.java
@@ -28,17 +28,22 @@ public final class LauncherActivity extends Activity {
     @Override
     public void onCreate(Bundle info) {
         super.onCreate(info);
-
         Intent intent = new Intent(this, LauncherService.class);
         startService(intent);
 
         Log.i(TAG, "Created launcher activity");
     }
 
+    @Override
+    protected void onResume() {
+        super.onResume();
+	Intent intent = new Intent(LauncherActivity.this, AppViewActivity.class);
+	startActivity(intent);
+    }
+
     @Override
     public void onDestroy() {
         Log.i(TAG, "Destroyed launcher activity");
-
         Intent intent = new Intent(this, LauncherService.class);
         stopService(intent);
 
