diff -uNr a/anbox/android/appmgr/Android.mk b/anbox/android/appmgr/Android.mk
--- a/anbox/android/appmgr/Android.mk	2020-04-12 22:56:14.991740643 +0800
+++ b/anbox/android/appmgr/Android.mk	2020-04-12 22:56:49.771740659 +0800
@@ -14,6 +14,5 @@
 	Home \
 	Launcher2 \
 	Launcher3 \
-	LatinIME \
 	QuickSearchBox
 include $(BUILD_PACKAGE)
diff -uNr a/anbox/android/opengl/system/egl/egl.cpp b/anbox/android/opengl/system/egl/egl.cpp
--- a/anbox/android/opengl/system/egl/egl.cpp	2020-04-12 22:56:14.999740643 +0800
+++ b/anbox/android/opengl/system/egl/egl.cpp	2020-04-12 22:58:54.799740716 +0800
@@ -463,6 +463,11 @@
     DEFINE_AND_VALIDATE_HOST_CONNECTION(EGL_FALSE);
 
     rcEnc->rcFlushWindowColorBuffer(rcEnc, rcSurface);
+	if (!buffer) {
+		ALOGE("queue null buffer to Surface");
+		setErrorReturn(EGL_BAD_ALLOC, EGL_FALSE);
+	}
+
     nativeWindow->queueBuffer_DEPRECATED(nativeWindow, buffer);
 
     if (nativeWindow->dequeueBuffer_DEPRECATED(nativeWindow, &buffer)) {
diff -uNr a/anbox/products/anbox.xml b/anbox/products/anbox.xml
--- a/anbox/products/anbox.xml	2020-04-12 22:56:15.055740643 +0800
+++ b/anbox/products/anbox.xml	2020-04-12 22:56:49.771740659 +0800
@@ -23,12 +23,15 @@
     <unavailable-feature name="android.software.managed_users" />
 
     <!-- Disable support for hardware we don't support -->
-    <unavailable-feature name="android.hardware.usb.accessory" />
-    <unavailable-feature name="android.hardware.usb.host" />
+    <feature name="android.hardware.usb.accessory" />
+    <feature name="android.hardware.usb.host" />
     <unavailable-feature name="android.hardware.wifi" />
     <unavailable-feature name="android.hardware.bluetooth" />
     <unavailable-feature name="android.hardware.bluetooth_le" />
     <unavailable-feature name="android.hardware.sensor.compass" />
     <unavailable-feature name="android.hardware.sensor.accelerometer" />
+    <feature name="android.software.app_widgets" />
+    <feature name="android.software.home_screen" />
+    <feature name="android.software.input_methods" />
 </permissions>
 
diff -uNr a/anbox/products/arm64/BoardConfig.mk b/anbox/products/arm64/BoardConfig.mk
--- a/anbox/products/arm64/BoardConfig.mk	2020-04-12 22:56:15.055740643 +0800
+++ b/anbox/products/arm64/BoardConfig.mk	2020-04-12 22:56:49.771740659 +0800
@@ -41,6 +41,10 @@
 # of an SDK AVD. Note that this operation only works on Linux for now
 ifeq ($(HOST_OS),linux)
 WITH_DEXPREOPT ?= true
+    ifneq ($(TARGET_BUILD_VARIANT),user)
+        # Retain classes.dex in APK's for non-user builds
+        DEX_PREOPT_DEFAULT := nostripping
+    endif
 endif
 
 # PDK does not use ext4 image, but it is added here to prevent build break.
diff -uNr a/anbox/src/anbox/rpc/message_processor.cpp b/anbox/src/anbox/rpc/message_processor.cpp
--- a/anbox/src/anbox/rpc/message_processor.cpp	2020-04-12 22:56:15.063740643 +0800
+++ b/anbox/src/anbox/rpc/message_processor.cpp	2020-04-12 22:56:49.771740659 +0800
@@ -45,7 +45,7 @@
 bool MessageProcessor::process_data(const std::vector<std::uint8_t> &data) {
   for (const auto &byte : data) buffer_.push_back(byte);
 
-  while (buffer_.size() > 0) {
+  while (buffer_.size() > 3) {
     const auto high = buffer_[0];
     const auto medium = buffer_[1];
     const auto low = buffer_[2];
diff -uNr a/anbox/tests/CMakeLists.txt b/anbox/tests/CMakeLists.txt
--- a/anbox/tests/CMakeLists.txt	2020-04-12 22:56:15.067740643 +0800
+++ b/anbox/tests/CMakeLists.txt	2020-04-12 22:56:49.771740659 +0800
@@ -19,8 +19,8 @@
 
     anbox-core
 
-    ${GMOCK_LIBRARIES}
     ${GTEST_BOTH_LIBRARIES}
+    ${GMOCK_LIBRARIES}
 
     ${ARGN}
 
